drivers: uart: nrfx: Support custom baudrates

bamkrs opened the pull request:
Nordics UARTE does support custom (non-standard) baudrates. The baudrate register controlls a 20bit baudrate generator. This patch adds a simple approximation which yields a very small error to generate custom baudrates. The default baudrates haven't been touched, thus the reccomended values stated in the reference are used.

Signed-off-by: Benedikt-Alexander Mokroß [github@bamkrs.de](mailto:github@bamkrs.de)

bamkrs committed:
drivers: uart: nrfx: Support custom baudrates

Nordics UARTE does support custom (non-standard) baudrates.
The baudrate register controlls a 20bit baudrate generator.
This patch adds a simple approximation which yields a very
small error to generate custom baudrates.
The default baudrates haven't been touched, thus the
recommended values stated in the reference are used.

Signed-off-by: Benedikt-Alexander Mokroß <github@bamkrs.de>

nika-nordic commented:
> Nordics UARTE does support custom (non-standard) baudrates.

I am not sure if that is true - inspecting Product Specification for various devices (e.g. [nRF52840](https://infocenter.nordicsemi.com/topic/ps_nrf52840/uarte.html?cp=5_0_0_5_33_8_26#register.BAUDRATE) ) shows that there are very specific baudrate settings that are supported and any violation cannot be guaranteed to work

bamkrs commented:
@nika-nordic I tested these on nRF52840 and nRF52833. Both yield my desired custom baudrates with an error below 2%.
If neccessary, I can modify my patch to only target nRF52840 and nRF52833 since I only tested it on these devices.

bamkrs commented:
@nika-nordic The standard baudrates are just a fixed counter values, too. Why should custom baudrates be subject to temperature drifts, but not those fixed values?

Further, Zephyr also adds support for baudrates [300](https://github.com/zephyrproject-rtos/zephyr/blob/zephyr-v3.3.0/drivers/serial/uart_nrfx_uarte.c#L353) and [600](https://github.com/zephyrproject-rtos/zephyr/blob/zephyr-v3.3.0/drivers/serial/uart_nrfx_uarte.c#L357) which both are not defined by the nordic SDK nor specified in the nRF528xx reference and exactly match the results yielded by the calculation.

| Baudrate | Calculated |  Library   | Error |
|----------|------------|------------|-------|
|      300 |      81920 |      - | 0.00% |
|      600 |     159744 |     - | 0.00% |
|     1200 |     323584 |     323584 | 0.00% |
|     2400 |     643072 |     643072 | 0.00% |
|     4800 |    1290240 |    1290240 | 0.00% |
|     9600 |    2576384 |    2576384 | 0.00% |
|    14400 |    3866624 |    3862528 | 0.11% |
|    19200 |    5152768 |    5152768 | 0.00% |
|    28800 |    7729152 |    7716864 | 0.16% |
|    31250 |    8388608 |    8388608 | 0.00% |
|    38400 |   10309632 |   10289152 | 0.20% |
|    56000 |   15032320 |   15007744 | 0.16% |
|    57600 |   15462400 |   15400960 | 0.40% |
|    76800 |   20615168 |   20615168 | 0.00% |
|   115200 |   30924800 |   30801920 | 0.40% |
|   230400 |   61845504 |   61865984 | 0.03% |
|   250000 |   67108864 |   67108864 | 0.00% |
|   460800 |  123695104 |  121634816 | 1.67% |
|   921600 |  247390208 |  251658240 | 1.73% |
|  1000000 |  268435456 |  268435456 | 0.00% |

When we run the calculation and use the actual baudrates stated in the nRF52840 reference, the error is even smaller and pretty close to those in the datasheet.

| Act.Baud | Calculated |  Library   | Error |
|----------|------------|------------|-------|
|      300 |      81920 |      - | 0.00% |
|      600 |     159744 |     - | 0.00% |
|     1205 |     323584 |     323584 | 0.00% |
|     2396 |     643072 |     643072 | 0.00% |
|     4808 |    1290240 |    1290240 | 0.00% |
|     9598 |    2576384 |    2576384 | 0.00% |
|    14401 |    3866624 |    3862528 | 0.11% |
|    19208 |    5156864 |    5152768 | 0.08% |
|    28777 |    7725056 |    7716864 | 0.11% |
|    31250 |    8388608 |    8388608 | 0.00% |
|    38369 |   10301440 |   10289152 | 0.12% |
|    55944 |   15015936 |   15007744 | 0.05% |
|    57554 |   15450112 |   15400960 | 0.32% |
|    76923 |   20647936 |   20615168 | 0.16% |
|   115108 |   30900224 |   30801920 | 0.32% |
|   231884 |   62246912 |   61865984 | 0.61% |
|   250000 |   67108864 |   67108864 | 0.00% |
|   457143 |  122712064 |  121634816 | 0.88% |
|   941176 |  252645376 |  251658240 | 0.39% |
|  1000000 |  268435456 |  268435456 | 0.00% |

(Values quickly calculated using https://onlinegdb.com/DN_hx22Dp)

MaureenHelm commented:
@anangl can you follow up please?

bamkrs committed:
drivers: uart: nrfx: Hide custom baudrates behind kconfig

Custom baudrates are not offically supported by Nordic.
However, they hint its OK to use the approximation for nRF52 UARTE
peripherals but recommend testing by the user.
This Patch hides the support for custom baudrates behind a kconfig option
forcing the user to explicitly it.
The kconfig option states a warning that custom baudrates should be
verified by the user.

Signed-off-by: Benedikt-Alexander Mokroß <github@bamkrs.de>

MaureenHelm commented:
@anangl please revisit

bamkrs commented:
@anangl Any input?
